h2o-core/src/main/java/water/DKVManager.java

```diff
@@ -17,14 +17,48 @@
    *
    * @param retainedKeys Keys of {@link Frame}s and {@link Model}s to be retained. Only Frame and Model keys are accepted.
    */
-  public static void retain(final Key[] retainedKeys){
+  public static void retain(final Key... retainedKeys) {
     final Set<Key> retainedSet = new HashSet<>(retainedKeys.length);
     retainedSet.addAll(Arrays.asList(retainedKeys));
     // Frames and models have multiple nested keys. Those must be extracted and kept from deletion as well.
     extractNestedKeys(retainedSet);
```

Comment:
Why do we need to do this for Models? If someone wants to keep the training Frame used to train a Model - it should be in "retainedKeys"

Reply:
I believe we don't need to do this at all now.

Reply:
It was assumed the model as a whole should be retained as functional. It is debatable whether the training frame should be part of it, but validation/xval models will be deleted as well if we don't extract them. I'm not sure whether users would expect such behavior.

On the other hand, they can always extract the keys themselves, making other layers responsible, giving other more control over the behavior of this method.

**I've decided removed this behavior for Models completely.**, as you propose. This means users will have to extract model's internal keys themselves, otherwise the internal sub-models won't survive.

Reply:
I demonstrated this behavior in `testCrossValidationModelsNotRetained` test.